### DISCUSSION TITLE
How does HSR follow given paths?

### DISCUSSION THREAD

**Discussion by yhiroya** [HumanSupportRobot/QandA#38] [Q&A]
I used another program to generate a path for the robot with (x,y,angle) as shown in the figure, and when I gave the generated path to HSR, HSR did not follow exactly on that path, but moved by rotating to reach the given angle at each node on the path even if it deviated from the path significantly.
What algorithm does HSR use for path following?
![image (7)](https://user-images.githubusercontent.com/114554742/211751177-7761ecd1-d2e6-4254-91ea-0d0b156566fa.png)

**Comment by hsr-admin** [comment]
When path is input directly to base_path_follower, the process is as follows

- Spline interpolation of the path

- Finds the nearest point of the robot's own position to the path after spline interpolation

- The command velocity in the tangential direction of the nearest point is calculated, and the velocity is output by adding a correction term proportional to the difference in position and angle between the nearest point and the robot's own position.

Therefore, if the path is spline-interpolated, the robot will take a trajectory that first moves forward once and then returns to the front.

This may cause the path to take a shortcut due to the calculation of the nearest point.


If you want the path to follow the specified point, it is recommended to input the path in the form of JointTrajectory.
By the way, could you tell me what you entered for path, is it base_path_follower?

**Reply by yhiroya** [reply]
Yes. We input our path for "base_path_follower" node.
However, we did not send it as ROS Topic.
We made actionlib client for "path_follow_action".
We assigned path data "PathFollowActionGoal.path" and send it using send_goal method.
Is there any difference in behavior between sending with Topic and sending with Actionlib's GOAL?
